File: factory-method-pattern/shape-creator/shape-creator.ts
interface Shape {
  draw(): string;
}

// <=========  Concrete products  ===========>
class Circle implements Shape {
  draw(): string {
    return "Drawing a circle";
  }
}

class Square implements Shape {
  draw(): string {
    return "Drawing a square :||";
  }
}

// <---------- Abstract creator ------------>
interface ShapeFactory {
  createShape(): Shape;
}

// --- Concrete creators ----
class CircleFactory implements ShapeFactory {
  createShape(): Shape {
    return new Circle();
  }
}

class SquareFactory implements ShapeFactory {
  createShape(): Shape {
    return new Square();
  }
}

function drawShape(factory: ShapeFactory) {
  const shape = factory.createShape();
  console.log(shape.draw());
}

drawShape(new CircleFactory());
drawShape(new SquareFactory());

File: factory-method-pattern/shape-creator/shape-creator.rs
// Abstract Product
trait Shape {
    fn draw(&self) -> String;
}

// Concrete Product 1
struct Circle;

impl Shape for Circle {
    fn draw(&self) -> String {
        "Drawing a circle".to_string()
    }
}

// Concrete Product 2
struct Square;

impl Shape for Square {
    fn draw(&self) -> String {
        "Drawing a square".to_string()
    }
}

// Abstract Creator
trait ShapeFactory {
    fn create_shape(&self) -> Box<dyn Shape>;
}

// Concrete Creator 1
struct CircleFactory;

impl ShapeFactory for CircleFactory {
    fn create_shape(&self) -> Box<dyn Shape> {
        Box::new(Circle)
    }
}

// Concrete Creator 2
struct SquareFactory;

impl ShapeFactory for SquareFactory {
    fn create_shape(&self) -> Box<dyn Shape> {
        Box::new(Square)
    }
}

fn draw_shape(fact: &dyn ShapeFactory) {
    let shape = fact.create_shape();
    println!("{}", shape.draw());
}

fn main() {
    let circle_factory = CircleFactory;
    let square_factory = SquareFactory;

    draw_shape(&circle_factory);
    draw_shape(&square_factory);
}
